### PR TITLE
Fix segfault on empty contract compiled w/ --error-recovery

### LINKED ISSUE #7296: --error-recovery segv's when there is no contract
## Description

Consder this invalid program

```solidity
pragma 0.5.11;
c
```

when run with `--error-recovery` this SEGV's: 

```shell
$ ./solc/solc --error-recovery --ast-json ../tmp/c.sol
Segmentation fault (core dumped)
```

## Environment

- Compiler version: 0.5.11 or current devlop
- Operating system: ubunto

## Steps to Reproduce

```solidity
pragma 0.5.11;
c
```

See  https://github.com/ethereum/solidity/pull/7293 for a possible fix.

### PR DESCRIPTION
**Pull request by erak**
Fixes https://github.com/ethereum/solidity/issues/7296.

### CODE REVIEW

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -281,53 +281,55 @@ bool CompilerStack::analyze()
 	{
```
Comment (leonardoalt) [inline libsolidity/interface/CompilerStack.cpp:281]: If `source->ast == nullptr` it seems like this entire block will do nothing.
Wouldn't it be better to just stop earlier if so?

**Comment by axic** [comment]
Nitpick: `with` is usually shortened to `w/` (where it makes sense, and not `/w` :wink:)

**Comment by rocky** [comment]
Something that's been in the back of my mind that's related to this is that there should be an option to do extended error recover _without_ ast generation. 

This would allow future bugs of this sort not to block use of use of extended error recovery. And there may be cases where you want extended error recovery without the AST generation.

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (erak) [reply libsolidity/interface/CompilerStack.cpp:281]: Mmh, wouldn't that solution need to check _all_ ASTs beforehand? 
```
for (Source const* source: m_sourceOrder)
    if (!source->ast)
        return false;
```
Maybe for some source files it's fine and we could still run analysis on them?

Comment (leonardoalt) [reply libsolidity/interface/CompilerStack.cpp:281]: Ah right, I missed the `for` loops ;)

File: libsolidity/interface/CompilerStack.cpp [hunk 2]
```diff
@@ -281,53 +281,55 @@ bool CompilerStack::analyze()
...
 		SyntaxChecker syntaxChecker(m_errorReporter, m_optimiserSettings.runYulOptimiser);
 		for (Source const* source: m_sourceOrder)
-			if (!syntaxChecker.checkSyntax(*source->ast))
+			if (source->ast &&!syntaxChecker.checkSyntax(*source->ast))
```
Comment (leonardoalt) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
			if (source->ast && !syntaxChecker.checkSyntax(*source->ast))
```
Below also

Comment (erak) [reply libsolidity/interface/CompilerStack.cpp]: Fixed all of them.

**Comment by erak** [comment]
Updated.

**Comment by erak** [comment]
@rocky What do you mean by _extended_ error recovery? In which aspects is that different to what's currently implemented?

**Comment by rocky** [comment]
@erak it has been a while since I looked at this. When a programmer uses the `--error-recovery` switch, two things happen:

1. the Steve C Johnson-style error recovery kicks in. This is a parser thing.
2. (partial) AST generation can happen when there are errors

However 2 may SEGV, while 1 never has. 

Suppose I have a front-end tool that makes use of the additional error message. If the program SEGVs then the tool doesn't get those additional messages because solc dies first.

A developer is then tempted to turn off getting the additional error messages to avoid the solc SEGVs, even though the tool can benefit in having those additional error messages.

In doing the experimental VSCode extension I ran into this and even mentioned it. See https://www.youtube.com/watch?v=lhQ0gO9Jzx8&t=9s.  

The SEGV that this addresses been in there for 3 months (and a couple of releases). 
Who knows if this will be the last SEGV problem as a result of partial AST generation? Separating additional parser error recovery from parital AST generation is therefore, I think, beneficial.

**Review by chriseth — APPROVED** [review]

**Pull request merged**